### PR TITLE
feat: moses and guacamol baselines training pipelines

### PR DESCRIPTION
**Pull request by Ashish13898**

### CODE REVIEW

File: src/gt4sd/domains/materials/scorer.py [hunk 1]
```diff
@@ -327,8 +371,24 @@ def score(self, smiles: str) -> float:
         Returns:
             A score for the given SMILES
         """
+        print(Chem.QED.qed(Chem.MolFromSmiles(smiles)))
```
Comment (drugilsberg) [inline src/gt4sd/domains/materials/scorer.py:374]: remove print

File: src/gt4sd/domains/materials/scorer.py [hunk 2]
```diff
@@ -216,6 +216,17 @@ def score(self, smiles: str) -> float:
         )
         return scoring_function.score_mol(Chem.MolFromSmiles(smiles))
 
+    def score_list(self, smiles_list: List[str]) -> List[float]:
```
Comment (drugilsberg) [inline src/gt4sd/domains/materials/scorer.py:219]: why not defining this in `TargetValueScorer` so we avoid repeating the same in every child class.

File: src/gt4sd/training_pipelines/moses/core.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+class MosesSavingArguments(TrainingPipelineArguments):
+    """Saving arguments related to PaccMann trainer."""
+
+    __name__ = "saving_args"
+
+    model_path: str = field(
+        metadata={"help": "Path where the model artifacts are stored."}
+    )
+    config_path: str = field(
+        metadata={"help": "Path where the config artifacts are stored."}
```
Comment (drugilsberg) [inline src/gt4sd/training_pipelines/moses/core.py:63]: `Path to the configuration file`?

File: src/gt4sd/training_pipelines/moses/organ/core.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        self,
+        model_args: Dict[str, Any],
+        common_args: Dict[str, Any],
+    ) -> None:  # type: ignore
+        params = {**common_args, **model_args}
+        parser = argparse.ArgumentParser()
+        for k, v in params.items():
+            parser.add_argument("--" + k, default=v)
+        args = parser.parse_known_args()[0]
+        print(vars(args))
```
Comment (drugilsberg) [inline src/gt4sd/training_pipelines/moses/organ/core.py:29]: remove print

File: src/gt4sd/training_pipelines/tests/test_training_guacamol_baselines_smiles_lstm_hc.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        "valid_every": 1000,
+        "n_epochs": 10,
+        "hidden_size": 512,
+        "n_layers": 2,
+        "rnn_dropout": 0.2,
+        "lr": 1e-3,
+    },
+    "training_args": {
+        "max_len": 100,
+        "output_dir": "/Users/ashishdave/Desktop/GT4SD/gt4sd-core/",
```
Comment (drugilsberg) [inline src/gt4sd/training_pipelines/tests/test_training_guacamol_baselines_smiles_lstm_hc.py:24]: remove path to dev environment

File: src/gt4sd/training_pipelines/tests/test_training_moses_organ.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+"""Moses Organ Trainer unit tests."""
```
Comment (drugilsberg) [inline src/gt4sd/training_pipelines/tests/test_training_moses_organ.py:1]: ORGAN?

File: src/gt4sd/training_pipelines/tests/test_training_moses_vae.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        "lr_start": 3 * 1e-4,
+        "lr_n_period": 10,
+        "lr_n_restarts": 10,
+        "lr_n_mult": 1,
+        "lr_end": 3 * 1e-4,
+        "n_last": 1000,
+        "n_jobs": 1,
+        "n_workers": 1,
+    },
+    "common_args": {
```
Comment (drugilsberg) [inline src/gt4sd/training_pipelines/tests/test_training_moses_vae.py:44]: not sure I follow the logic, can't we use the same ones we use for the paccmann-vae tests?

**Review by drugilsberg — CHANGES_REQUESTED** [review]

File: src/gt4sd/training_pipelines/moses/core.py (hunk 3 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/moses/core.py:63]: So for Moses we have to pass [vocab, config and model path](https://github.com/GT4SD/moses/blob/main/moses/script_utils.py#:~:text=common_arg.add_argument(%27%2D%2Dconfig_load,load%20the%20config%27)) and while training it's updating vocab and config as well as creating model artifacts therefore I added config and vocab to the saving arguments as well. 

Should we just save the model artifacts instead?

File: src/gt4sd/domains/materials/scorer.py (hunk 2 above)
Comment (Ashish13898) [reply src/gt4sd/domains/materials/scorer.py:219]: Just noticed it was already defined, I was not using it directly from parent, I will remove these methods from the child classes, thanks :)

File: src/gt4sd/training_pipelines/moses/core.py (hunk 3 above)
Comment (drugilsberg) [reply src/gt4sd/training_pipelines/moses/core.py:63]: We have to save them all, sorry for the comment it's coming from my limited knowledge of the pipeline.

Comment (Ashish13898) [reply src/gt4sd/training_pipelines/moses/core.py:63]: No worries at all :)

File: src/gt4sd/training_pipelines/tests/test_training_guacamol_baselines_smiles_lstm_hc.py (hunk 5 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/tests/test_training_guacamol_baselines_smiles_lstm_hc.py]: I have pushed a change for this, thanks :)

File: src/gt4sd/training_pipelines/moses/organ/core.py (hunk 4 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/moses/organ/core.py]: I have pushed a change for this, thanks :)

File: src/gt4sd/domains/materials/scorer.py (hunk 1 above)
Comment (Ashish13898) [reply src/gt4sd/domains/materials/scorer.py]: I have pushed a change for this, thanks :)

File: src/gt4sd/training_pipelines/tests/test_training_moses_organ.py (hunk 6 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/tests/test_training_moses_organ.py:1]: Apologies if I understand this comment wrong but did you meant that we should change Organ to ORGAN?

Comment (drugilsberg) [reply src/gt4sd/training_pipelines/tests/test_training_moses_organ.py:1]: yep :)

File: src/gt4sd/training_pipelines/tests/test_training_moses_vae.py (hunk 7 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/tests/test_training_moses_vae.py:49]: I have pushed a change for this, thanks :)

File: src/gt4sd/training_pipelines/tests/test_training_moses_organ.py (hunk 6 above)
Comment (Ashish13898) [reply src/gt4sd/training_pipelines/tests/test_training_moses_organ.py]: I have pushed a change for this, thanks :)

**Comment by drugilsberg** [comment]
TODOs left:

- [ ] Moses Organ trainer causes issue when run via CLI
- [ ] Check MOSES fork to make sure script_utils work fine
- [ ] double check logging in the CLI.

**Comment by drugilsberg** [comment]
> > 
> 
> Not investigating Organ trainer further removing the pipeline.

Trying the exact configuration from test worked, enabling organ pipeline.

**Comment by drugilsberg** [comment]
Parameters tested:

```
GuacaMol LSTM

gt4sd-trainer --training_pipeline_name guacamol-lstm-trainer --train_smiles_filepath src/gt4sd/training_pipelines/tests/molecules.smiles --test_smiles_filepath src/gt4sd/training_pipelines/tests/molecules.smiles --output_dir /tmp/test-guacamol


gt4sd-saving --training_pipeline_name guacamol-lstm-trainer --algorithm_application SMILESLSTMHCGenerator --target_version testing-trainer-v1 --model_filepath /tmp/test-guacamol/model_final_2.464.pt --model_config_filepath /tmp/test-guacamol/model_final_2.464.json

{"n_epochs": 1}

gt4sd-inference --algorithm_application SMILESLSTMHCGenerator --algorithm_version testing-trainer-v1 --number_of_samples 2 --target '{"isomer_scorer": {"target": 5.0, "target_smile": "NCCCCC"}}' --configuration_file /tmp/guacamol-hc-config.json


gt4sd-saving --training_pipeline_name guacamol-lstm-trainer --algorithm_application SMILESLSTMPPOGenerator --target_version testing-trainer-v1 --model_filepath /tmp/test-guacamol/model_final_2.464.pt --model_config_filepath /tmp/test-guacamol/model_final_2.464.json

{"num_epochs": 1, "episode_size": 4, "optimize_batch_size": 4}

gt4sd-inference --algorithm_application SMILESLSTMPPOGenerator --algorithm_version testing-trainer-v1 --number_of_samples 2 --target '{"isomer_scorer": {"target": 5.0, "target_smile": "NCCCCC"}}' --configuration_file /tmp/guacamol-ppo-config.json


Moses Organ

{
        "embedding_size": 32,
        "hidden_size": 512,
        "num_layers": 2,
        "dropout": 0.0,
        "discriminator_layers": "[(100, 1), (200, 2), (200, 3), (200, 4), (200, 5), (100, 6), (100, 7), (100, 8), (100, 9), (100, 10), (160, 15), (160, 20)]",
        "discriminator_dropout": 0.0,
        "reward_weight": 0.7,
        "n_jobs": 1,
        "generator_pretrain_epochs": 1,
        "discriminator_pretrain_epochs": 1,
        "pg_iters": 1,
        "n_batch": 4,
        "lr": 1e-4,
        "n_workers": 1,
        "max_length": 50,
        "clip_grad": 5,
        "rollouts": 16,
        "generator_updates": 1,
        "discriminator_updates": 1,
        "discriminator_epochs": 1,
        "n_ref_subsample": 1,
        "addition_rewards": "sa,qed",
        "seed": 0,
        "device": "cpu",
        "save_frequency": 1,
        "model_save": "/tmp/moses-organ-trainer/model.pt",
        "log_file": "/tmp/moses-organ-trainer/log.txt",
        "config_save": "/tmp/moses-organ-trainer/config.pt",
        "vocab_save": "/tmp/moses-organ-trainer/vocab.pt",
        "train_load": "src/gt4sd/training_pipelines/tests/molecules.smiles",
        "val_load": "src/gt4sd/training_pipelines/tests/molecules.smiles"
}

gt4sd-trainer --training_pipeline_name moses-organ-trainer --configuration_file /tmp/test-organ-trainer.json

gt4sd-saving --training_pipeline_name moses-organ-trainer --algorithm_application OrganGenerator --target_version testing-trainer-v1 --model_path /tmp/moses-organ-trainer/model.pt --config_path /tmp/moses-organ-trainer/config.pt --vocab_path /tmp/moses-organ-trainer/vocab.pt

gt4sd-inference --algorithm_application OrganGenerator --algorithm_version testing-trainer-v1 --number_of_samples 4

Moses VAE

gt4sd-trainer --training_pipeline_name moses-vae-trainer --model_save /tmp/moses-trainer/model.pt --log_file /tmp/moses-trainer/log.txt --config_save /tmp/moses-trainer/config.pt --vocab_save /tmp/moses-trainer/vocab.pt --train_load src/gt4sd/training_pipelines/tests/molecules.smiles --val_load src/gt4sd/training_pipelines/tests/molecules.smiles --lr_n_restarts 1 --lr_n_mult 1 --lr_n_period 1

gt4sd-saving --training_pipeline_name moses-vae-trainer --algorithm_application VaeGenerator --target_version testing-trainer-v1 --model_path /tmp/moses-trainer/model.pt --config_path /tmp/moses-trainer/config.pt --vocab_path /tmp/moses-trainer/vocab.pt

gt4sd-inference --algorithm_application VaeGenerator --algorithm_version testing-trainer-v1 --number_of_samples 4
```

**Review by drugilsberg — APPROVED** [review]

**Pull request merged**